lib/maths/common/unittest/CLinearAlgebraTest.cc

```diff
@@ -1159,12 +1160,12 @@ BOOST_AUTO_TEST_CASE(testShims) {
     {
         BOOST_TEST_REQUIRE(TMatrix4(0.0) ==
                            maths::common::las::conformableZeroMatrix(vector1));
-        BOOST_TEST_REQUIRE(TMatrix(4, 0.0) ==
-                           maths::common::las::conformableZeroMatrix(vector2));
-        BOOST_TEST_REQUIRE(TDenseMatrix::Zero(4, 4) ==
-                           maths::common::las::conformableZeroMatrix(vector3));
-        BOOST_TEST_REQUIRE(TDenseMatrix::Zero(4, 4) ==
-                           maths::common::las::conformableZeroMatrix(vector4));
+        BOOST_TEST_REQUIRE((TMatrix(4, 0.0) ==
+                           maths::common::las::conformableZeroMatrix(vector2)));
+        BOOST_TEST_REQUIRE((TDenseMatrix::Zero(4, 4) ==
+                           maths::common::las::conformableZeroMatrix(vector3)));
+        BOOST_TEST_REQUIRE((TDenseMatrix::Zero(4, 4) ==
+                           maths::common::las::conformableZeroMatrix(vector4)));
```

Comment:
This suggests that `conformableZeroMatrix` used to be printable but now isn't.

Why is that?

Reply:
It has more to do with the overloading of the `==` operator. Here's the compilation error message:

```
In file included from CLinearAlgebraTest.cc:24:
In file included from /Users/eds/src/elasticsearch/ml-cpp/include/test/BoostTestCloseAbsolute.h:15:
In file included from /usr/local/include/boost-1_77/boost/test/test_tools.hpp:55:
In file included from /usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:18:
/usr/local/include/boost-1_77/boost/test/utils/is_forward_iterable.hpp:216:60: error: no type named 'value_type' in 'std::iterator_traits<void>'
    typedef typename std::iterator_traits<const_iterator>::value_type value_type;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:446:30: note: in instantiation of template class 'boost::unit_test::bt_iterator_traits<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>, true>' requested here
BOOST_TEST_FOR_EACH_COMP_OP( DEFINE_COLLECTION_COMPARISON )
                             ^
/usr/local/include/boost-1_77/boost/test/tools/assertion.hpp:340:79: note: in instantiation of template class 'boost::test_tools::assertion::op::EQ<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>, ml::maths::common::CDenseMatrix<double>>' requested here
class binary_expr : public expression_base<binary_expr<LExpr,Rhs,OP>,typename OP::result_type> {
                                                                              ^
CLinearAlgebraTest.cc:1164:53: note: in instantiation of template class 'boost::test_tools::assertion::binary_expr<boost::test_tools::assertion::value_expr<const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>>, ml::maths::common::CDenseMatrix<double>, boost::test_tools::assertion::op::EQ<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>, ml::maths::common::CDenseMatrix<double>>>' requested here
        BOOST_TEST_REQUIRE(TDenseMatrix::Zero(4, 4) ==
                                                    ^
In file included from CLinearAlgebraTest.cc:24:
In file included from /Users/eds/src/elasticsearch/ml-cpp/include/test/BoostTestCloseAbsolute.h:15:
In file included from /usr/local/include/boost-1_77/boost/test/test_tools.hpp:55:
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:446:30: error: no matching function for call to 'compare_collections'
BOOST_TEST_FOR_EACH_COMP_OP( DEFINE_COLLECTION_COMPARISON )
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/boost-1_77/boost/test/tools/assertion.hpp:72:5: note: expanded from macro 'BOOST_TEST_FOR_EACH_COMP_OP'
    action( ==, EQ, !=, NE )                \
    ^~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:427:16: note: expanded from macro 'DEFINE_COLLECTION_COMPARISON'
        return assertion::op::compare_collections( lhs, rhs,        \
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/boost-1_77/boost/test/tools/assertion.hpp:365:20: note: in instantiation of member function 'boost::test_tools::assertion::op::EQ<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>, ml::maths::common::CDenseMatrix<double>>::eval' requested here
        return OP::eval( m_lhs.value(), m_rhs );
                   ^
/usr/local/include/boost-1_77/boost/test/tools/assertion.hpp:374:42: note: in instantiation of member function 'boost::test_tools::assertion::binary_expr<boost::test_tools::assertion::value_expr<const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>>, ml::maths::common::CDenseMatrix<double>, boost::test_tools::assertion::op::EQ<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>, ml::maths::common::CDenseMatrix<double>>>::value' requested here
        assertion_result const expr_res( value() );
                                         ^
CLinearAlgebraTest.cc:1166:9: note: in instantiation of member function 'boost::test_tools::assertion::binary_expr<boost::test_tools::assertion::value_expr<const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>>, ml::maths::common::CDenseMatrix<double>, boost::test_tools::assertion::op::EQ<Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, Eigen::Matrix<double, -1, -1, 0>>, ml::maths::common::CDenseMatrix<double>>>::evaluate' requested here
        BOOST_TEST_REQUIRE(TDenseMatrix::Zero(4, 4) ==
        ^
/usr/local/include/boost-1_77/boost/test/tools/interface.hpp:152:8: note: expanded from macro 'BOOST_TEST_REQUIRE'
    2, BOOST_TEST_TOOL_UNIV, BOOST_TEST_TOOL_UNIV_EX, REQUIRE, __VA_ARGS__ )\
       ^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:344:1: note: candidate function template not viable: no known conversion from 'bool_<false>' to 'bool_<true>' for 4th argument
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<OP>*, mpl::true_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:299:1: note: candidate template ignored: could not match 'EQ<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::EQ<L, R> >*, mpl::true_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:308:1: note: candidate template ignored: could not match 'EQ<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::EQ<L, R> >*, mpl::false_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:317:1: note: candidate template ignored: could not match 'NE<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::NE<L, R> >*, mpl::true_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:326:1: note: candidate template ignored: could not match 'NE<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::NE<L, R> >*, mpl::false_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:353:1: note: candidate template ignored: could not match 'LT<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::LT<L, R> >*, mpl::false_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:362:1: note: candidate template ignored: could not match 'LE<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::LE<L, R> >*, mpl::false_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:371:1: note: candidate template ignored: could not match 'GT<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::GT<L, R> >*, mpl::false_ )
^
/usr/local/include/boost-1_77/boost/test/tools/collection_comparison_op.hpp:380:1: note: candidate template ignored: could not match 'GE<type-parameter-0-2, type-parameter-0-3>' against 'int'
compare_collections( Lhs const& lhs, Rhs const& rhs, boost::type<op::GE<L, R> >*, mpl::false_ )
^
```

Reply:
Yes, but I think the reason Boost overloads `==` is so that it can print out the lhs and rhs if the test fails.  So the reason the overloading fails is because one side isn't printable.  So some type somewhere has changed from something that's printable to something that isn't.

I will look back to the original cppunit -> Boost.Test PR to see if there are any clues.

Reply:
Thanks!

Reply:
There were no clues in #720.

It seems that some Eigen iterator is now missing a `value_type`.  This is probably the result of some change to get rid of C++17 deprecations 🤦 .

Still changing these assertions is not too bad.  If they ever fail whoever debugs it will just have to print out the values in some other way.